third-party/realdds/src/dds-device-server.cpp

```diff
@@ -4,264 +4,91 @@
-        _notifications_loop.start();
-        _active = true;
-    }
-
-    ~dds_notifications_server()
-    {
-        // Mark this class as inactive and wake up the active object do we can properly stop it.
-        _active = false;
-        _send_notification_cv.notify_all();
-        _notifications_loop.stop();
-
-        if( nullptr != _data_writer )
-        {
-            DDS_API_CALL_NO_THROW( _publisher->delete_datawriter( _data_writer ) );
-        }
-
-        if( nullptr != _topic )
-        {
-            DDS_API_CALL_NO_THROW( _participant->get()->delete_topic( _topic ) );
-        }
-    };
-
-    void send_notification( topics::raw::device::notification&& notification ) 
-    {
-        std::unique_lock< std::mutex > lock( _notification_send_mutex );
-        if( ! _instant_notifications.enqueue( std::move( notification ) ) )
-        {
-            LOG_ERROR( "error while trying to enqueue a message id:"
-                       << notification.id() << " to instant notifications queue" );
-        }
-    };
-
-    void add_init_notification( topics::raw::device::notification && msg ) 
-    {
-        std::unique_lock< std::mutex > lock( _notification_send_mutex );
-        topics::raw::device::notification msg_to_move( msg );
-        _init_notifications.push_back( std::move( msg_to_move ) );
-    };
-    
 
-private:
-
-    void send_init_msgs()
-    {
-        // Send all initialization notifications
-        for( auto notification : _init_notifications )
-        {
-            DDS_API_CALL( _data_writer->write( &notification ) );
-        }
-    }
-
-    std::shared_ptr< dds_participant > _participant;
-    eprosima::fastdds::dds::Publisher * _publisher;
-    eprosima::fastdds::dds::Topic * _topic;
-    eprosima::fastdds::dds::DataWriter * _data_writer;
-    dds_notifications_client_listener _clients_listener;
-    active_object<> _notifications_loop;
-    single_consumer_queue< topics::raw::device::notification > _instant_notifications;
-    std::vector< topics::raw::device::notification > _init_notifications;
-    std::mutex _notification_send_mutex;
-    std::condition_variable _send_notification_cv;
-    std::atomic_bool _send_init_msgs = { false }; 
-    std::atomic_bool _new_instant_notification = { false };  
-    std::atomic_bool _active = { false };
-};
-
-//-----------------------------------------------------------------------------------------------------//
-// DDS device server is in charge of handling the device control, streams and notifications
-//-----------------------------------------------------------------------------------------------------//
 dds_device_server::dds_device_server( std::shared_ptr< dds_participant > const & participant,
                                       const std::string & topic_root )
     : _participant( participant )
-    , _publisher( nullptr )
+    , _publisher( std::make_shared< dds_publisher >( participant ))
```

Comment:
Why is `_publisher` initialized here and other members (e.g. `_notification_server`) in `init`?

Reply:
The publisher holds the participant (`_participant` has already been removed in the next PR).
But otherwise there is no specific reason -- it was just an invariant with the life of the server so it was created in the ctor, And it can do some checking while at it (to create it, the participant may not be null). There's no real reason to delay it as I see no use-case where you'd have a server that's not going to be initialized...